File: src/Sdks/SdkLoader.fs
module Helium.Sdks.SdkLoader
open Newtonsoft.Json
open System.IO
open System.Threading.Tasks
open FSharp.Control.Tasks.V2
open Helium.Util


let private createSerSettings () =
    let settings = JsonSerializerSettings()
    settings.Converters.Add <| Converters.StringEnumConverter()
    settings.Formatting <- Formatting.Indented
    settings

let loadSdk (sdkFile: string): SdkInfo Task = task {
    let! data = File.ReadAllTextAsync(sdkFile)
    return JsonConvert.DeserializeObject<SdkInfo>(data, createSerSettings())
}

let private findFiles dir =
    Directory.EnumerateFiles(dir, "*.json", SearchOption.AllDirectories)

let loadSdks =
    findFiles >> Seq.map loadSdk
    
let saveSdk (sdk: SdkInfo) (path: string): Task =
    File.WriteAllTextAsync(path, JsonConvert.SerializeObject(sdk, typeof<SdkInfo>, createSerSettings()))
    
let sdkSha256 (sdk: SdkInfo): string =
    let settings = createSerSettings()
    settings.Formatting <- Formatting.None
    JsonConvert.SerializeObject(sdk, typeof<SdkInfo>, settings) |> HashUtil.Sha256UTF8
    

File: src/Sdks/SdkLoader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Helium.Util;
using Newtonsoft.Json;

namespace Helium.Sdks
{
    public static class SdkLoader
    {
        private static JsonSerializerSettings SerSettings => new JsonSerializerSettings {
            Formatting = Formatting.Indented,
        };

        public static async Task<SdkInfo> LoadSdk(string sdkFile) {
            var data = await File.ReadAllTextAsync(sdkFile);
            return JsonConvert.DeserializeObject<SdkInfo>(data, SerSettings) ?? throw new Exception("Could not load sdk.");
        }

        public static async IAsyncEnumerable<SdkInfo> LoadSdks(string dir) {
            foreach(var file in Directory.EnumerateFiles(dir, "*.json", SearchOption.AllDirectories)) {
                yield return await LoadSdk(file);
            }
        }

        public static Task SaveSdk(string sdkFile, SdkInfo sdk) =>
            File.WriteAllTextAsync(sdkFile, JsonConvert.SerializeObject(sdk, typeof(SdkInfo), SerSettings));

        public static string SdkSha256(SdkInfo sdk) {
            var settings = SerSettings;
            settings.Formatting = Formatting.None;
            var json = JsonConvert.SerializeObject(sdk, typeof(SdkInfo), settings);
            return HashUtil.Sha256UTF8(json);
        }
    }
}